blog/_posts/2022-04-26-secret-management-the-soft-way.md

```diff
@@ -0,0 +1,212 @@
+  target:
+    name: k8s-secret  # name of the k8s Secret to be created
+  data:
+  - secretKey: FOO  # name of the GCPSM secret key
+    remoteRef:
+      key: FOO
+  - secretKey: BAR  # name of the GCPSM secret key
+    remoteRef:
+      key: BAR
+```
+
+
+## HashiCorp Vault & External Secrets Operator
+The External Secrets Operator does not only connect to secret management products provided by the big cloud vendors. If you have bigger security concerns than the mere storing of secrets, you might be using something like [Hashicorp Vault][vault] to control not only your secrets, but also they way they are created and managed.
+If you’re using ESO already, all you have to do is allow kubernetes to access your vault by creating a role for it. You can make this role as granular as you like, for example, only allowing access to specific secrets or paths.
+In the secret store we then point to the kubernetes authentication secret stored in Vault.
+```bash
+$ vault policy write eso-policy -<<EOF     
+path "kv/data/vault-secret"                                                  
+{  capabilities = ["read"]                
+}                         
+EOF
+
+$ vault write auth/kubernetes/role/eso-role \
+    bound_service_account_names=external-secrets \
+    bound_service_account_namespaces=es \
+    policies=eso-policy \
+    ttl=24h
+```
+```yaml
+apiVersion: external-secrets.io/v1alpha1
+kind: ClusterSecretStore
+metadata:
+  name: vault-backend
+spec:
+  provider:
+    vault:
+      server: "http://vault.vault:8200"
+      path: "kv"
+      version: "v2"
+      auth:
+        kubernetes:
+          mountPath: "kubernetes"
+          role: "eso-role"
+```
+(There's a bit more to it than shown here. For a full guide check out [this tutorial][eso-vault].)
+
+The External Resource will look almost exactly the same as in the example above.
+```yaml
+apiVersion: external-secrets.io/v1alpha1
+kind: ExternalSecret
+metadata:
+  name: vault-external-secret
+spec:
+  secretStoreRef:
+    name: vault-backend
+    kind: ClusterSecretStore
+  target:
+    name: k8s-secret
+  data:
+  - secretKey: FOO
+    remoteRef:
+      key: vault-secret
+      property: FOO
+  - secretKey: SOURCE
+    remoteRef:
+      key: vault-secret
+      property: SOURCE
+```
+Of course the big plus in this scenario is the full control over the entire lifecycle of a secret, however it comes with its own challenges. Managing a tool like vault does not only require pure operational effort, but also intentionally creating proper security procedures to fully leverage its usefulness.
+
+---
+
+To summarize: There are a multitude of ways to get secrets inside Kubernetes, but keep in mind that you need to manage and secure them outside the cluster. Safeguarding your secrets relies on every single employee. Inconvenient or convoluted processes will compromise security efforts, as people will look for the path of the least resistance and grow less alert over time.
```

Comment:
(as well as secure your cluster) -- I feel like I want to read this again, but up to you how much that point needs to be stressed.